lib/extended_config.rb

```diff
@@ -0,0 +1,48 @@
+require 'yaml'
+require 'fileutils'
+
+class ExtendedConfig
+  IAAS_TYPE_KEY = 'IAAS_TYPE'.freeze
+  DEFAULT_IAAS_TYPE = 'openstack'.freeze
+
+  attr_reader :extended_config
+  # private_class_method :new
+
+  def default_config
+    {
+      'default' => {
+        'iaas' => default_iaas_type
+      }
+    }
+  end
+
+  def initialize(env = {})
+    @extended_config = env
+  end
+
+  def ==(other)
+    return false unless other.is_a?(ExtendedConfig)
+    @extended_config == other.extended_config
+  end
+
+  private
+
+  def default_iaas_type
+    @extended_config.fetch(IAAS_TYPE_KEY, DEFAULT_IAAS_TYPE)
+  end
+
+  class ExtendedConfigBuilder
+    def initialize
+      @current_config = {}
+    end
+
+    def build
+      ExtendedConfig.new(@current_config)
+    end
+
+    def iaas_type(name)
```

Comment:
The name of this method induces us to believe that it's a getter more than a setter method. I'd rather call it "with_iaas_type" especially since it returns self.